### PR TITLE
feat(sumologicextension): fail for 4xx errors except 429

### PR DESCRIPTION
**Pull request by sumo-drosiek**
Signed-off-by: Dominik Rosiek <drosiek@sumologic.com>

### CODE REVIEW

File: pkg/extension/sumologicextension/extension.go [hunk 1]
```diff
@@ -286,10 +288,20 @@ func (se *SumologicExtension) registerCollector(ctx context.Context, collectorNa
 			zap.Int("status_code", res.StatusCode),
 			zap.String("response", buff.String()),
 		)
-		return CollectorCredentials{}, fmt.Errorf(
-			"failed to register the collector, got HTTP status code: %d",
-			res.StatusCode,
-		)
+
+		// Return unrecoverable error for 4xx status codes except 429
+		if res.StatusCode >= 400 && res.StatusCode < 500 && res.StatusCode != 429 {
```
Comment (pmm-sumo) [inline pkg/extension/sumologicextension/extension.go:293]: We might consider reusing this abstraction: https://github.com/open-telemetry/opentelemetry-collector/blob/1fada742275967c44c7c01445f097793b9d0caae/consumer/consumererror/permanent.go

Comment (sumo-drosiek) [reply pkg/extension/sumologicextension/extension.go:291]: Could you check it now? as abstraction is not exported I use backoff.Pernament

File: pkg/extension/sumologicextension/extension.go [hunk 2]
```diff
@@ -286,10 +286,19 @@ func (se *SumologicExtension) registerCollector(ctx context.Context, collectorNa
...
+			return CollectorCredentials{}, backoff.Permanent(fmt.Errorf(
+				"failed to register the collector, got HTTP status code: %d",
```
Comment (astencel-sumo) [inline pkg/extension/sumologicextension/extension.go:293]: Not strictly related to this very PR, but aren't we losing the original error's message and only passing the response code here? Wouldn't the original error's message potentially contain additional info for the user, helping them fix the problem?

File: pkg/extension/sumologicextension/extension.go [hunk 3]
```diff
@@ -322,8 +331,9 @@ func (se *SumologicExtension) registerCollectorWithBackoff(ctx context.Context,
 		se.logger.Warn("Collector registration failed: ", zap.Error(err))
 
 		nbo := se.backOff.NextBackOff()
-		if nbo == se.backOff.Stop {
-			return CollectorCredentials{}, fmt.Errorf("collector registration failed")
+		// Return error if backoff reaches the limit or uncoverable error is spotted
+		if _, ok := err.(*backoff.PermanentError); nbo == se.backOff.Stop || ok {
```
Comment (astencel-sumo) [inline pkg/extension/sumologicextension/extension.go:335]: This `ok` variable name gives absolutely no information on what it means and makes this line incomprehensible.

A name like `errorIsPermanent` would make this line much easier to read and understand.

**Review by astencel-sumo — APPROVED** [review]
Added some comments, but not blockers to going ahead with it 👍

File: pkg/extension/sumologicextension/extension.go (hunk 3 above)
Comment (sumo-drosiek) [reply pkg/extension/sumologicextension/extension.go:335]: `ok` is used for type casting in golang, so I would live it

File: pkg/extension/sumologicextension/extension.go (hunk 2 above)
Comment (sumo-drosiek) [reply pkg/extension/sumologicextension/extension.go:293]: Response is logged above
```
		se.logger.Debug("Collector registration failed",
			zap.Int("status_code", res.StatusCode),
			zap.String("response", buff.String()),
		)
```

**Review by pmm-sumo — APPROVED** [review]

**Pull request merged**